From 3c973d8c6122f042ed3015165c8d2a3ee05d37ae Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Thu, 21 Dec 2023 14:38:16 +0800
Subject: [PATCH 1/4] avce: Use VDEnc for YUY2/AYUV/RGB formats

TGLx+ supports YUY2/AYUV/RGB formats, but VDEnc is required. [1]

This makes the command below works on TGL

gst-launch-1.0 videotestsrc ! video/x-raw,format=YUY2 ! msdkh264enc ! fakesink

[1] https://github.com/intel/media-driver/blob/master/docs/media_features.md#media-features-summary

OSV: https://jira.devtools.intel.com/browse/PKT-19566
Signed-off-by: Haihao Xiang <haihao.xiang@intel.com>
---
 .../mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp b/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp
index 5c8c7e20..b5dfcfaa 100644
--- a/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp
+++ b/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp
@@ -1439,9 +1439,13 @@ MFX_ERR_NONE - if no errors
 mfxStatus MfxHwH264Encode::SetLowPowerDefault(MfxVideoParam& par, const eMFXHWType& platform)
 {
     mfxStatus sts = CheckTriStateOption(par.mfx.LowPower) ? MFX_ERR_NONE : MFX_WRN_INCOMPATIBLE_VIDEO_PARAM;
+    auto fcc = par.mfx.FrameInfo.FourCC;
 
-    if (!H264ECaps::IsVmeSupported(platform))
-    {   // DualPipe (aka VME) is not available
+    if (!H264ECaps::IsVmeSupported(platform) ||
+        (fcc == MFX_FOURCC_YUY2 ||
+         fcc == MFX_FOURCC_AYUV ||
+         fcc == MFX_FOURCC_RGB4))
+    {   // DualPipe (aka VME) is not available or YUY2/AYUV/RGB input formats
         par.mfx.LowPower = MFX_CODINGOPTION_ON;
         return sts;
     }
-- 
2.43.0

